resource-recorder/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+use async_trait::async_trait;
+use dal::{Dal, DalError, Resource};
+use prost_types::TimestampError;
+use shuttle_common::{backends::auth::VerifyClaim, claims::Scope};
+use shuttle_proto::resource_recorder::{
+    self, resource_recorder_server::ResourceRecorder, ProjectResourcesRequest, RecordRequest,
+    ResourcesResponse, ResultResponse, ServiceResourcesRequest,
+};
+use thiserror::Error;
+use tonic::{Request, Response, Status};
+
+pub mod args;
+mod dal;
+mod r#type;
+
+pub use dal::Sqlite;
+use tracing::error;
+use ulid::DecodeError;
+
+/// A wrapper to capture any error possible with this service
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("could not decode id: {0}")]
+    UlidDecode(#[from] DecodeError),
+
+    #[error("failed to interact with database: {0}")]
+    Dal(#[from] DalError),
+
+    #[error("could not parse resource type: {0}")]
+    String(String),
+
+    #[error("could not parse timestamp: {0}")]
+    Timestamp(#[from] TimestampError),
+}
+
+// thiserror is not happy to handle a `#[from] String`
+impl From<String> for Error {
+    fn from(value: String) -> Self {
+        Self::String(value)
+    }
+}
+
+pub struct Service<D> {
+    dal: D,
+}
+
+impl<D> Service<D>
+where
+    D: Dal + Send + Sync + 'static,
+{
+    pub fn new(dal: D) -> Self {
+        Self { dal }
+    }
+
+    /// Record the addition of a new resource
+    async fn add(&self, request: RecordRequest) -> Result<(), Error> {
+        self.dal
+            .add_resources(
+                request.project_id.parse()?,
+                request.service_id.parse()?,
+                request
+                    .resources
+                    .into_iter()
+                    .map(TryInto::<Resource>::try_into)
+                    .collect::<Result<_, _>>()?,
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    /// Get the resources that belong to a project
+    async fn project_resources(
+        &self,
+        project_id: String,
+    ) -> Result<Vec<resource_recorder::Resource>, Error> {
+        let resources = self.dal.get_project_resources(project_id.parse()?).await?;
+
+        Ok(resources.into_iter().map(Into::into).collect())
+    }
+
+    /// Get the resources that belong to a service
+    async fn service_resources(
+        &self,
+        service_id: String,
+    ) -> Result<Vec<resource_recorder::Resource>, Error> {
+        let resources = self.dal.get_service_resources(service_id.parse()?).await?;
+
+        Ok(resources.into_iter().map(Into::into).collect())
+    }
+
+    /// Delete a resource
+    async fn delete_resource(&self, resource: resource_recorder::Resource) -> Result<(), Error> {
+        self.dal.delete_resource(&resource.try_into()?).await?;
+
+        Ok(())
+    }
```

Comment:
Yeh, wondering if this is enough. Should we cleanup our infra as well? I would cover this in another PR probably, but the endpoint shouldn't be exposed yet given we remove the resources from the persistence and we might be left with dangling resources in the infra, but maybe I am missing something? Also, we should factor in this endpoint in the custom resources definitions too (and infra that lives outside our management e.g. Turso).

Reply:
Yip, this should not be exposed yet. Whatever will call this (deployer) will have to also call provisioner to cleanup the infra.

Reply:
Do we still want to expose this RPC (even internally)? Maybe we can add a clarifying comment on the RPC method implementation, that this handler still lacks the infrastructure teardown. What do you think?

Reply:
Nah, the infrastructure teardown won't go through this RPC anyway (something else will do the teardown and just record it here). And trying to add the comment here feels out of place.